// @dydxprotocol/perpetual/dist/src/lib/Constants.js
"use strict";
/*

    Copyright 2020 dYdX Trading Inc.

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.

*/
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var bignumber_js_1 = __importDefault(require("bignumber.js"));
var types_1 = require("./types");
var ONE_MINUTE_IN_SECONDS = new bignumber_js_1.default(60);
var ONE_HOUR_IN_SECONDS = ONE_MINUTE_IN_SECONDS.times(60);
var ONE_DAY_IN_SECONDS = ONE_HOUR_IN_SECONDS.times(24);
var ONE_YEAR_IN_SECONDS = ONE_DAY_IN_SECONDS.times(365);
exports.PRICES = {
    NONE: new types_1.Price(0),
    ONE: new types_1.Price(1),
};
exports.FEES = {
    ZERO: new types_1.Fee(0),
    ONE_BIP: new types_1.Fee('1e-4'),
    ONE_PERCENT: new types_1.Fee('1e-2'),
};
exports.INTEGERS = {
    ONE_MINUTE_IN_SECONDS: ONE_MINUTE_IN_SECONDS,
    ONE_HOUR_IN_SECONDS: ONE_HOUR_IN_SECONDS,
    ONE_DAY_IN_SECONDS: ONE_DAY_IN_SECONDS,
    ONE_YEAR_IN_SECONDS: ONE_YEAR_IN_SECONDS,
    ZERO: new bignumber_js_1.default(0),
    ONE: new bignumber_js_1.default(1),
    ONES_255: new bignumber_js_1.default('115792089237316195423570985008687907853269984665640564039457584007913129639935'),
};
exports.ADDRESSES = {
    ZERO: '0x0000000000000000000000000000000000000000',
    TEST: [
        '0x06012c8cf97bead5deae237070f9587f8e7a266d',
        '0x22012c8cf97bead5deae237070f9587f8e7a266d',
        '0x33012c8cf97bead5deae237070f9587f8e7a266d',
        '0x44012c8cf97bead5deae237070f9587f8e7a266d',
        '0x55012c8cf97bead5deae237070f9587f8e7a266d',
        '0x66012c8cf97bead5deae237070f9587f8e7a266d',
        '0x77012c8cf97bead5deae237070f9587f8e7a266d',
        '0x88012c8cf97bead5deae237070f9587f8e7a266d',
        '0x99012c8cf97bead5deae237070f9587f8e7a266d',
        '0xaa012c8cf97bead5deae237070f9587f8e7a266d',
    ],
};
// ============ P1TraderConstants.sol ============
exports.TRADER_FLAG_ORDERS = new bignumber_js_1.default(1);
exports.TRADER_FLAG_LIQUIDATION = new bignumber_js_1.default(2);
exports.TRADER_FLAG_DELEVERAGING = new bignumber_js_1.default(4);
// ============ P1Orders.sol ============
exports.ORDER_FLAGS = {
    IS_BUY: 1,
    IS_DECREASE_ONLY: 2,
    IS_NEGATIVE_LIMIT_FEE: 4,
};
// ============ P1FundingOracle.sol ============
// Rate limiting is based on a 55 minute period, equal to the funding rate update interval
// of one hour, with five minutes as a buffer.
var FUNDING_LIMIT_PERIOD = exports.INTEGERS.ONE_MINUTE_IN_SECONDS.times(55);
// Funding rate limits set by the smart contract.
exports.FUNDING_RATE_MAX_ABS_VALUE = types_1.FundingRate.fromEightHourRate('0.0075').roundedDown();
exports.FUNDING_RATE_MAX_ABS_DIFF_PER_UPDATE = exports.FUNDING_RATE_MAX_ABS_VALUE;
exports.FUNDING_RATE_MAX_ABS_DIFF_PER_SECOND = exports.FUNDING_RATE_MAX_ABS_VALUE.div(FUNDING_LIMIT_PERIOD).roundedDown();
