// @dydxprotocol/perpetual/dist/src/modules/PriceOracle.js
"use strict";
/*

    Copyright 2020 dYdX Trading Inc.

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.

*/
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var types_1 = require("../lib/types");
var PriceOracle = /** @class */ (function () {
    function PriceOracle(contracts) {
        this.contracts = contracts;
        this.oracles = {};
    }
    Object.defineProperty(PriceOracle.prototype, "address", {
        // ============ Getter Functions ============
        get: function () {
            return this.contracts.p1MakerOracle.options.address;
        },
        enumerable: true,
        configurable: true
    });
    PriceOracle.prototype.getPrice = function (options) {
        if (options === void 0) { options = {}; }
        return __awaiter(this, void 0, void 0, function () {
            var combinedOptions, oracle, price;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        combinedOptions = __assign({ from: this.contracts.perpetualProxy.options.address }, options);
                        return [4 /*yield*/, this.getCurrentOracleContract(options)];
                    case 1:
                        oracle = _a.sent();
                        return [4 /*yield*/, this.contracts.call(oracle.methods.getPrice(), combinedOptions)];
                    case 2:
                        price = _a.sent();
                        return [2 /*return*/, types_1.Price.fromSolidity(price)];
                }
            });
        });
    };
    PriceOracle.prototype.getRoute = function (sender, options) {
        if (options === void 0) { options = {}; }
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.contracts.call(this.contracts.p1MakerOracle.methods._ROUTER_(sender), options)];
            });
        });
    };
    PriceOracle.prototype.getOracleAdjustment = function (oracleAddress, options) {
        if (options === void 0) { options = {}; }
        return __awaiter(this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.contracts.call(this.contracts.p1MakerOracle.methods._ADJUSTMENTS_(oracleAddress), options)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, types_1.BaseValue.fromSolidity(result)];
                }
            });
        });
    };
    // ============ Admin Functions ============
    PriceOracle.prototype.setRoute = function (sender, oracle, options) {
        if (options === void 0) { options = {}; }
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.contracts.send(this.contracts.p1MakerOracle.methods.setRoute(sender, oracle), options)];
            });
        });
    };
    PriceOracle.prototype.setAdjustment = function (oracle, adjustment, options) {
        if (options === void 0) { options = {}; }
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.contracts.send(this.contracts.p1MakerOracle.methods.setAdjustment(oracle, adjustment.toSolidity()), options)];
            });
        });
    };
    // ============ Helper Functions ============
    PriceOracle.prototype.getCurrentOracleContract = function (options) {
        if (options === void 0) { options = {}; }
        return __awaiter(this, void 0, void 0, function () {
            var oracleAddress;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.contracts.call(this.contracts.perpetualV1.methods.getOracleContract(), options)];
                    case 1:
                        oracleAddress = _a.sent();
                        return [2 /*return*/, this.getOracle(oracleAddress)];
                }
            });
        });
    };
    PriceOracle.prototype.getOracle = function (oracleAddress) {
        if (this.oracles[oracleAddress]) {
            return this.oracles[oracleAddress];
        }
        var contract = this.contracts.p1MakerOracle.clone();
        contract.options.address = oracleAddress;
        this.oracles[oracleAddress] = contract;
        return contract;
    };
    return PriceOracle;
}());
exports.PriceOracle = PriceOracle;
